node/src/keyshare/mod.rs

```diff
@@ -1,34 +1,76 @@
+/// This structure is used in context, where we should have all key shares present (e.g. "Running" state)
 #[derive(Clone, Serialize, Deserialize)]
 pub struct RootKeyshareData {
     pub epoch: u64,
-    pub private_share: Scalar,
-    pub public_key: AffinePoint,
+    pub ecdsa: <EcdsaSignatureProvider as SignatureProvider>::KeygenOutput,
+}
+
+/// A helper structure that allows loading each share separately.
+/// It's useful when we want to introduce a new signature provider without breaking back-compatibility.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct PartialRootKeyshareData {
+    pub epoch: u64,
+    pub ecdsa: Option<<EcdsaSignatureProvider as SignatureProvider>::KeygenOutput>,
+}
+
+impl PartialRootKeyshareData {
+    /// Returns a complete RootKeyshareData if all fields are set.
+    pub(crate) fn as_complete(&self) -> Option<RootKeyshareData> {
+        let ecdsa = self.ecdsa.clone()?;
+        Some(RootKeyshareData {
+            epoch: self.epoch,
+            ecdsa,
+        })
+    }
 }
 
 impl RootKeyshareData {
-    pub fn keygen_output(&self) -> KeygenOutput<Secp256k1> {
-        KeygenOutput {
-            private_share: self.private_share,
-            public_key: self.public_key,
-        }
+    pub fn new(
+        epoch: u64,
+        ecdsa: <EcdsaSignatureProvider as SignatureProvider>::KeygenOutput,
+    ) -> Self {
+        Self { epoch, ecdsa }
     }
 
-    pub fn new(epoch: u64, keygen_output: KeygenOutput<Secp256k1>) -> Self {
-        Self {
-            epoch,
-            private_share: keygen_output.private_share,
-            public_key: keygen_output.public_key,
-        }
+    /// If some data exists already, check that we can safely overwrite it
+    fn compare_against_existing_share(
+        to_save: &RootKeyshareData,
+        existing: &PartialRootKeyshareData,
+    ) -> anyhow::Result<()> {
+        match existing.as_complete() {
+            // Key shares for all signature providers generated -> we do resharing
+            //  -> we have to check, that epoch is greater than it was.
+            Some(existing) => {
+                if existing.epoch >= to_save.epoch {
+                    return Err(anyhow::anyhow!(
+                        "Refusing to overwrite existing keyshare of epoch {} with new keyshare of older epoch {}",
+                        existing.epoch,
+                        to_save.epoch,
+                    ));
+                }
+            }
+
+            // We are still generating a key share for some signature provider -> we have to make sure,
+            //  that we do not overwrite existing key shares
+            None => {
+                if let Some(ecdsa) = &existing.ecdsa {
+                    if ecdsa.private_share != to_save.ecdsa.private_share {
+                        return Err(anyhow::anyhow!(
+                            "Refusing to overwrite existing ecdsa keyshare"
+                        ));
+                    }
+                }
```

Comment:
I couldn't think of a way how to enforce invariants here, i.e. how would a user adding a new keyshare know, that this code needs to be adjusted?